alf/experience_replayers/replay_buffer.py

```diff
@@ -286,49 +331,65 @@ def get_batch(self, batch_size, batch_length):
                                             self._buffer)
 
-            if self._prioritized_sampling:
-                indices = self._env_id_idx_to_index(env_ids,
-                                                    self.circular(start_pos))
-                avg_weight = self._sum_tree.nnz / self._sum_tree.summary()
-                info = info._replace(
-                    importance_weights=self._sum_tree[indices] * avg_weight)
-
             if self._postprocess_exp_fn:
                 result, info = self._postprocess_exp_fn(self, result, info)
 
         return convert_device(result), convert_device(info)
 
+    def _recent_sample(self, batch_size, batch_length):
+        return self._sample(batch_size, batch_length, self._recent_data_steps)
+
     def _uniform_sample(self, batch_size, batch_length):
         min_size = self._current_size.min()
         assert min_size >= batch_length, (
             "Not all environments have enough data. The smallest data "
             "size is: %s Try storing more data before calling get_batch" %
             min_size)
+        return self._sample(batch_size, batch_length)
 
+    def _sample(self, batch_size, batch_length, recent_data_steps=None):
         batch_size_per_env = batch_size // self._num_envs
         remaining = batch_size % self._num_envs
-        if batch_size_per_env > 0:
-            env_ids = torch.arange(self._num_envs).repeat(batch_size_per_env)
-        else:
-            env_ids = torch.zeros(0, dtype=torch.int64)
+        env_ids = torch.arange(self._num_envs).repeat(batch_size_per_env)
         if remaining > 0:
-            eids = torch.randperm(self._num_envs)[:remaining]
-            env_ids = torch.cat([env_ids, eids], dim=0)
+            remaining_eids = torch.randperm(self._num_envs)[:remaining]
+            env_ids = torch.cat([env_ids, remaining_eids], dim=0)
 
         r = torch.rand(*env_ids.shape)
-
-        num_positions = self._current_size - batch_length + 1
-        num_positions = num_positions[env_ids]
-        pos = (r * num_positions).to(torch.int64)
-        pos += (self._current_pos - self._current_size)[env_ids]
+        if not self._with_replacement:
+            # For sampling without replacement,  we want r's for the same env to
+            # be roughly evenly spaced.
+            num_samples_per_env = batch_size_per_env * torch.ones(
+                (self._num_envs, ), dtype=torch.int64)
+            # Each i means that the corresponding sample is the i-th sample from
+            # that environment.
+            i = torch.arange(batch_size_per_env).unsqueeze(-1).expand(
+                -1, self._num_envs).reshape(-1)
+            if remaining > 0:
+                num_samples_per_env[remaining_eids] += 1
+                remaining_i = batch_size_per_env * torch.ones(
+                    (remaining, ), dtype=torch.int64)
+                i = torch.cat([i, remaining_i])
+
+            r = (i + r) / num_samples_per_env[env_ids]
```

Comment:
comment that this isn't really uniform sampling at the global level, but uniform sampling within each evenly spaced region?

Reply:
Fair enough.  It is uniform, just not independent.